File: ratio_calculate.cpp
#include <iostream>
#include <vector>
#include <string>
#include <memory>    // for std::unique_ptr
#include <stdexcept> // for std::runtime_error
#include <cmath>
#include <chrono>
#include <omp.h>
#include "fitsio.h"

constexpr int DATA_SIZE_X = 512;
constexpr int DATA_SIZE_Y = 512;
constexpr float CENTER_X = 256.0f;
constexpr float CENTER_Y = 256.0f;
constexpr int CELL_SIZE = 16; 
constexpr int GRID_W = DATA_SIZE_X / CELL_SIZE;
constexpr int GRID_H = DATA_SIZE_Y / CELL_SIZE;

constexpr int MAX_EVENTS_PER_CELL = 100; 
constexpr float PSF_RADIUS = 7.0f;       
constexpr int ITERATION = 5;             
constexpr float BKG_RATE = 1e-5f;        
constexpr float T_EXP = 1000.0f;         
constexpr float INV_BKG = 1.0f / BKG_RATE;
constexpr float INITIAL_R = 0.05f;

constexpr float THRESHOLD = 8.0f;      
constexpr int WINDOW_SIZE = 21;        
constexpr int W_RADIUS = WINDOW_SIZE / 2; 
constexpr int BORDER = 10;               

// Helper function to get 2D index from flattened 1D array
inline int idx(int x, int y) {
    return x * DATA_SIZE_Y + y;
}

struct Event {
    double x;
    double y;
};

struct Source {
    float x;
    float y;
    float rate;
};

void check_fits_status(int status) {
    if (status) {
        char err_text[FLEN_ERRMSG];
        fits_read_errmsg(err_text);
        throw std::runtime_error(std::string("CFITSIO Error: ") + err_text);
    }
}



/**
Performs the core likelihood calculation. This function ports both precompute_and_build_grid and core_calculation from the Python script.
data       Vector of all photon events.
ratio_grid Output pointer for the likelihood map (must be pre-allocated).
R_grid     Output pointer for the rate map (must be pre-allocated).
 */
void core_calculation(const std::vector<Event>& data, float* ratio_grid, float* R_grid) {
    
    // --- 1. Precomputation (from precompute_and_build_grid) ---
    
    // --- 1a. Geometry Maps ---
    // Allocate memory for PSF geometry maps
    auto s_maj_map_ptr = std::make_unique<float[]>(DATA_SIZE_X * DATA_SIZE_Y);
    auto s_min_map_ptr = std::make_unique<float[]>(DATA_SIZE_X * DATA_SIZE_Y);
    auto sin_map_ptr   = std::make_unique<float[]>(DATA_SIZE_X * DATA_SIZE_Y);
    auto cos_map_ptr   = std::make_unique<float[]>(DATA_SIZE_X * DATA_SIZE_Y);
    auto norm_map_ptr  = std::make_unique<float[]>(DATA_SIZE_X * DATA_SIZE_Y);
    
    // Get raw pointers for easier access
    float* s_maj = s_maj_map_ptr.get();
    float* s_min = s_min_map_ptr.get();
    float* sa    = sin_map_ptr.get();
    float* ca    = cos_map_ptr.get();
    float* nc    = norm_map_ptr.get();

    const float max_dist_inv = 1.0f / 362.039f; 
    const float pi = 3.1415926535f;

    // This loop is not parallelized in Python
    for (int x = 0; x < DATA_SIZE_X; ++x) {
        float dx = CENTER_X - x - 0.5f; 
        for (int y = 0; y < DATA_SIZE_Y; ++y) {
            float dy = CENTER_Y - y - 0.5f; 
            float dist = std::sqrt(dx * dx + dy * dy);
            
            float ratio = dist * max_dist_inv;
            float sigma_minor = 0.5f + ratio * 2.5f; 
            float eccentricity = 0.9f * ratio;     
            
            float ecc_sq = eccentricity * eccentricity;
            float denom = (ecc_sq >= 1.0f) ? 0.001f : std::sqrt(1.0f - ecc_sq); 
            float sigma_major = sigma_minor / denom;

            float angle = std::atan2(dy, dx); 
            
            int index = idx(x, y);
            s_maj[index] = sigma_major;
            s_min[index] = sigma_minor;
            sa[index]    = std::sin(angle);
            ca[index]    = std::cos(angle);
            nc[index]    = 1.0f / (2.0f * pi * sigma_major * sigma_minor); 
        }
    }

    // --- 1b. Grid Indexing ---
    auto grid_counts_ptr = std::make_unique<int[]>(GRID_W * GRID_H);
    // This is a 3D flattened array: grid[gx][gy][event_index]
    auto grid_indices_ptr = std::make_unique<int[]>(GRID_W * GRID_H * MAX_EVENTS_PER_CELL);
    
    int* g_cnt = grid_counts_ptr.get();
    int* g_idx = grid_indices_ptr.get();

    // Initialize grid counts to zero
    std::fill_n(g_cnt, GRID_W * GRID_H, 0);

    long n_events = data.size();
    for (int i = 0; i < n_events; ++i) {
        int gx = static_cast<int>(data[i].x) / CELL_SIZE; 
        int gy = static_cast<int>(data[i].y) / CELL_SIZE;
        
        if (gx >= 0 && gx < GRID_W && gy >= 0 && gy < GRID_H) {
            int grid_index = gx * GRID_H + gy; // 2D index for grid_counts
            int c = g_cnt[grid_index];
            if (c < MAX_EVENTS_PER_CELL) {
                // 3D index for grid_indices
                g_idx[grid_index * MAX_EVENTS_PER_CELL + c] = i;
                g_cnt[grid_index]++; 
            }
        }
    }
    
    // --- 2. Core Calculation (Parallelized) ---
    
    // Initialize output grids to zero
    std::fill_n(ratio_grid, DATA_SIZE_X * DATA_SIZE_Y, 0.0f);
    std::fill_n(R_grid, DATA_SIZE_X * DATA_SIZE_Y, 0.0f);

    // This is the C++/OpenMP equivalent of 'prange(GRID_W)'
    #pragma omp parallel for
    for (int gx = 0; gx < GRID_W; ++gx) {
        
        // This array is local to each thread, simulating the 'stack array'
        float local_s[256]; 

        for (int gy = 0; gy < GRID_H; ++gy) {
            
            // --- Active Grid Check ---
            //
            bool has_events = false;
            for (int ix = std::max(0, gx - 1); ix < std::min(GRID_W, gx + 2); ++ix) {
                for (int iy = std::max(0, gy - 1); iy < std::min(GRID_H, gy + 2); ++iy) {
                    if (g_cnt[ix * GRID_H + iy] > 0) {
                        has_events = true;
                        break;
                    }
                }
                if (has_events) break;
            }
            if (!has_events) continue; 

            // --- Pixel Loop ---
            //
            int x_start = gx * CELL_SIZE;
            int y_start = gy * CELL_SIZE;
            int x_end = std::min(x_start + CELL_SIZE, DATA_SIZE_X);
            int y_end = std::min(y_start + CELL_SIZE, DATA_SIZE_Y);

            for (int x = x_start; x < x_end; ++x) {
                for (int y = y_start; y < y_end; ++y) {
                    
                    int local_count = 0;
                    int p_idx = idx(x, y); // Pixel index
                    
                    // --- Gather Photons from 3x3 Grid ---
                    for (int ix = std::max(0, gx - 1); ix < std::min(GRID_W, gx + 2); ++ix) {
                        for (int iy = std::max(0, gy - 1); iy < std::min(GRID_H, gy + 2); ++iy) {
                            
                            int grid_index = ix * GRID_H + iy;
                            int cnt = g_cnt[grid_index];
                            if (cnt == 0) continue; 

                            for (int k = 0; k < cnt; ++k) {
                                int event_idx = g_idx[grid_index * MAX_EVENTS_PER_CELL + k];
                                float dx_ev = data[event_idx].x - x;
                                float dy_ev = data[event_idx].y - y;
                                
                                // Bounding Box Check
                                if (std::abs(dx_ev) <= PSF_RADIUS && std::abs(dy_ev) <= PSF_RADIUS) {
                                    // PSF Calc
                                    float t1 = (dx_ev * ca[p_idx] + dy_ev * sa[p_idx]) / s_maj[p_idx];
                                    float t2 = (-dx_ev * sa[p_idx] + dy_ev * ca[p_idx]) / s_min[p_idx];
                                    float expo = (t1 * t1 + t2 * t2) * 0.5f;
                                    float val = std::exp(-expo) * nc[p_idx];
                                    
                                    if (local_count < 256) {
                                        local_s[local_count] = val; 
                                        local_count++;
                                    }
                                }
                            }
                        }
                    } // end 3x3 grid search
                    
                    if (local_count == 0) continue; 

                    // --- R Iteration ---
                    float R = INITIAL_R;
                    for (int iter = 0; iter < ITERATION; ++iter) {
                        float temp_R = 0.0f;
                        for (int k = 0; k < local_count; ++k) {
                            float s = local_s[k];
                            temp_R += (R * s) / ((R * s + BKG_RATE) * T_EXP);
                        }
                        R = temp_R;
                    }
                    R_grid[p_idx] = R; 

                    // --- Log Likelihood ---
                    float log_sum = 0.0f;
                    for (int k = 0; k < local_count; ++k) {
                        log_sum += std::log(1.0f + R * local_s[k] * INV_BKG);
                    }
                    ratio_grid[p_idx] = log_sum - (T_EXP * R);

                } // end y
            } // end x
        } // end gy
    } // end gx (parallel loop)
}

std::vector<Source> find_local_maxima(const float* ratio_grid, const float* R_grid) {
    std::vector<Source> sources;

    // This loop can also be parallelized, as each pixel is independent
    // We add 'schedule(dynamic)' because sources are sparse, so some
    // threads might finish their rows much faster than others.
    #pragma omp parallel for schedule(dynamic)
    for (int r = BORDER; r < DATA_SIZE_X - BORDER; ++r) {
        for (int c = BORDER; c < DATA_SIZE_Y - BORDER; ++c) {
            
            int p_idx = idx(r, c); // Use our helper: r * DATA_SIZE_Y + c
            float val = ratio_grid[p_idx];

            // 1. Check threshold
            if (val < THRESHOLD) {
                continue;
            }

            // 2. Check if it's a local maximum
            bool is_max = true;
            for (int i = r - W_RADIUS; i <= r + W_RADIUS; ++i) {
                for (int j = c - W_RADIUS; j <= c + W_RADIUS; ++j) {
                    if (i == r && j == c) {
                        continue;
                    }
                    // Check if neighbor is greater
                    if (ratio_grid[idx(i, j)] > val) { 
                        is_max = false;
                        break;
                    }
                }
                if (!is_max) {
                    break;
                }
            }

            // 3. If it is, add to list
            if (is_max) {
                // We must use 'omp critical' to safely add to the
                // shared 'sources' vector from multiple threads.
                #pragma omp critical
                {
                    sources.push_back({static_cast<float>(r), 
                                       static_cast<float>(c), 
                                       R_grid[p_idx]});
                }
            }
        } // end for c
    } // end for r (parallel)

    return sources;
}

void write_fits_results(const std::string& filename, const std::vector<Source>& sources) {
    fitsfile* fptr = nullptr; // C-style file pointer
    int status = 0;   // CFITSIO status variable
    
    long n_sources = sources.size();
    if (n_sources == 0) {
        std::cout << "No sources to write, skipping FITS update." << std::endl;
        return;
    }

    try {
        // 1. Open the FITS file in READWRITE mode
        fits_open_file(&fptr, filename.c_str(), READWRITE, &status);
        check_fits_status(status);

        // 2. Move to the 2nd HDU (binary table)
        int hdutype;
        fits_movabs_hdu(fptr, 2, &hdutype, &status);
        check_fits_status(status);

        // 3. Get column numbers
        int x_col, y_col, rate_col;
        fits_get_colnum(fptr, CASEINSEN, (char*)"x", &x_col, &status);
        check_fits_status(status);
        fits_get_colnum(fptr, CASEINSEN, (char*)"y", &y_col, &status);
        check_fits_status(status);
        fits_get_colnum(fptr, CASEINSEN, (char*)"countrate", &rate_col, &status);
        check_fits_status(status);

        // 4. Prepare C-style arrays from the std::vector<Source>
        // must match the data types in the FITS file.
        // 'x' and 'y' are format 'I' (long integer in C)
        // 'countrate' is format 'D' (double in C)
        std::unique_ptr<long[]> x_data(new long[n_sources]);
        std::unique_ptr<long[]> y_data(new long[n_sources]);
        std::unique_ptr<double[]> rate_data(new double[n_sources]);

        for (long i = 0; i < n_sources; ++i) {
            // Cast our float results to the required FITS types
            x_data[i]    = static_cast<long>(sources[i].x);
            y_data[i]    = static_cast<long>(sources[i].y);
            rate_data[i] = static_cast<double>(sources[i].rate);
        }

        // 5. Write the data to the columns
        long firstrow = 1;
        long firstelem = 1;
        fits_write_col(fptr, TLONG,   x_col,    firstrow, firstelem, n_sources, x_data.get(),    &status);
        check_fits_status(status);
        fits_write_col(fptr, TLONG,   y_col,    firstrow, firstelem, n_sources, y_data.get(),    &status);
        check_fits_status(status);
        fits_write_col(fptr, TDOUBLE, rate_col, firstrow, firstelem, n_sources, rate_data.get(), &status);
        check_fits_status(status);

        // 6. Close the FITS file
        fits_close_file(fptr, &status);
        check_fits_status(status);
        
        std::cout << "Successfully wrote " << n_sources << " sources to " << filename << std::endl;

    } catch (const std::exception& e) {
        if (fptr) fits_close_file(fptr, &status);
        std::cerr << "Error writing to FITS file: " << e.what() << std::endl;
    }
}

std::vector<Event> read_fits_data(const std::string& filename) {
    fitsfile* fptr; 
    int status = 0;   
    
    try {
        
        fits_open_file(&fptr, filename.c_str(), READONLY, &status);
        check_fits_status(status);
       
        int hdutype;
        fits_movabs_hdu(fptr, 2, &hdutype, &status);
        check_fits_status(status);

        if (hdutype != BINARY_TBL) {
            throw std::runtime_error("FITS file does not contain a binary table at HDU 2.");
        }

        long nrows;
        fits_get_num_rows(fptr, &nrows, &status);
        check_fits_status(status);

       
        int x_col, y_col;
        fits_get_colnum(fptr, CASEINSEN, (char*)"x", &x_col, &status);
        check_fits_status(status);
        fits_get_colnum(fptr, CASEINSEN, (char*)"y", &y_col, &status);
        check_fits_status(status);
     
        std::unique_ptr<double[]> x_data(new double[nrows]);
        std::unique_ptr<double[]> y_data(new double[nrows]);
     
        long firstrow = 1;
        long firstelem = 1;
        double nulval = 0.0; 
        int anynul;      

        fits_read_col(fptr, TDOUBLE, x_col, firstrow, firstelem, nrows, &nulval, x_data.get(), &anynul, &status);
        check_fits_status(status);
        fits_read_col(fptr, TDOUBLE, y_col, firstrow, firstelem, nrows, &nulval, y_data.get(), &anynul, &status);
        check_fits_status(status);

        std::vector<Event> data;
        data.resize(nrows);
        for (long i = 0; i < nrows; ++i) {
            data[i].x = x_data[i];
            data[i].y = y_data[i];
        }

        fits_close_file(fptr, &status);
        check_fits_status(status);
        
        return data;

    } catch (const std::exception& e) {
        if (fptr) fits_close_file(fptr, &status);
        std::cerr << e.what() << std::endl;
        return std::vector<Event>();
    }
}

int main() {
    try {
        auto t0 = std::chrono::high_resolution_clock::now();

        // 1. Read Data
        std::cout << "Reading data..." << std::endl;
        std::vector<Event> data = read_fits_data("mock_data.fits");
        
        if (data.empty()) {
            std::cerr << "Failed to read data. Exiting." << std::endl;
            return 1;
        }
        std::cout << "Successfully read " << data.size() << " events." << std::endl;

        auto t1 = std::chrono::high_resolution_clock::now();
        std::cout << "Data IO time: " 
                  << std::chrono::duration<double>(t1 - t0).count() << "s" << std::endl;

        // 2. Allocate memory for grids
        auto ratio_grid_ptr = std::make_unique<float[]>(DATA_SIZE_X * DATA_SIZE_Y);
        auto R_grid_ptr     = std::make_unique<float[]>(DATA_SIZE_X * DATA_SIZE_Y);
        float* ratio_grid = ratio_grid_ptr.get();
        float* R_grid     = R_grid_ptr.get();

        // 3. Calculate
        std::cout << "Start calculating map..." << std::endl;
        core_calculation(data, ratio_grid, R_grid); 

        auto t2 = std::chrono::high_resolution_clock::now();
        std::cout << "Map calculation time: " 
                  << std::chrono::duration<double>(t2 - t1).count() << "s" << std::endl;

        // 4. Detect Sources (Stub)
        std::vector<Source> sources = find_local_maxima(ratio_grid, R_grid); 
        std::cout << "Detected sources (stub): " << sources.size() << std::endl;

        auto t3 = std::chrono::high_resolution_clock::now();
        std::cout << "Detection time (stub): " 
                  << std::chrono::duration<double>(t3 - t2).count() << "s" << std::endl;

        // 5. Write Results (Stub)
        write_fits_results("detection_info.fits", sources); 

        auto t4 = std::chrono::high_resolution_clock::now();
        std::cout << "Data IO time (stub): " 
                  << std::chrono::duration<double>(t4 - t3).count() << "s" << std::endl;
        
        std::cout << "Total time: " 
                  << std::chrono::duration<double>(t4 - t0).count() << "s" << std::endl;

    } catch (const std::exception& e) {
        std::cerr << "An unexpected error occurred: " << e.what() << std::endl;
        return 1;
    }
    return 0;
}




int __previous_main_for_testing_io() {
    try {
        auto t0 = std::chrono::high_resolution_clock::now();

        std::cout << "Attempting to read 'mock_data.fits'..." << std::endl;
        std::vector<Event> data = read_fits_data("mock_data.fits");

        auto t1 = std::chrono::high_resolution_clock::now();

        if (data.empty()) {
            std::cerr << "Failed to read data." << std::endl;
            return 1;
        }

        std::cout << "Successfully read " << data.size() << " events." << std::endl;
        std::cout << "Data IO time: " << std::chrono::duration<double>(t1 - t0).count() << "s" << std::endl;


        std::cout << "First 5 events:" << std::endl;
        for (int i = 0; i < 5 && i < data.size(); ++i) {
            std::cout << "Event " << i << ": (x=" << data[i].x << ", y=" << data[i].y << ")" << std::endl;
        }
        
    } catch (const std::exception& e) {
        std::cerr << "An unexpected error occurred: " << e.what() << std::endl;
        return 1;
    }

    return 0;
}

File: ratio_calculate.py
import numpy as np
from astropy.io import fits
import math
from numba import njit, prange
import time
DATA_SIZE_X = 512
DATA_SIZE_Y = 512
CENTER_X = 256.0
CENTER_Y = 256.0
# 网格大小设定为 16 (略大于 PSF_radius 7.5，保证只需查相邻格子)
CELL_SIZE = 16 
GRID_W = DATA_SIZE_X // CELL_SIZE
GRID_H = DATA_SIZE_Y // CELL_SIZE
MAX_EVENTS_PER_CELL = 100  # 假设每个格子最多这么多光子，防止溢出，可视情况调整

@njit(fastmath=True, cache=True)
def precompute_and_build_grid(data):
    # --- Geometry Map (Float32) ---
    # 这一步其实可以用 global memory 缓存，避免每次都算，但这里为了纯净算在耗时里
    # 简化：因为 cos/sin 对称，其实可以优化，但 Numba 这里不是瓶颈
    sigma_major_map = np.empty((DATA_SIZE_X, DATA_SIZE_Y), dtype=np.float32)
    sigma_minor_map = np.empty((DATA_SIZE_X, DATA_SIZE_Y), dtype=np.float32)
    sin_map = np.empty((DATA_SIZE_X, DATA_SIZE_Y), dtype=np.float32)
    cos_map = np.empty((DATA_SIZE_X, DATA_SIZE_Y), dtype=np.float32)
    norm_map = np.empty((DATA_SIZE_X, DATA_SIZE_Y), dtype=np.float32)

    max_dist_inv = 1.0 / 362.039

    for x in range(DATA_SIZE_X):
        dx = CENTER_X - x - 0.5
        for y in range(DATA_SIZE_Y):
            dy = CENTER_Y - y - 0.5
            dist = math.sqrt(dx*dx + dy*dy)
            
            ratio = dist * max_dist_inv
            sigma_minor = 0.5 + ratio * 2.5
            eccentricity = 0.9 * ratio
            
            ecc_sq = eccentricity * eccentricity
            denom = 0.001 if ecc_sq >= 1.0 else math.sqrt(1.0 - ecc_sq)
            sigma_major = sigma_minor / denom

            angle = math.atan2(dy, dx)
            
            sigma_major_map[x, y] = sigma_major
            sigma_minor_map[x, y] = sigma_minor
            sin_map[x, y] = math.sin(angle)
            cos_map[x, y] = math.cos(angle)
            norm_map[x, y] = 1.0 / (6.283185 * sigma_major * sigma_minor)

    # --- Grid Indexing ---
    n_events = len(data)
    grid_counts = np.zeros((GRID_W, GRID_H), dtype=np.int32)
    grid_indices = np.zeros((GRID_W, GRID_H, MAX_EVENTS_PER_CELL), dtype=np.int32)

    for i in range(n_events):
        ev_x = data[i, 0]
        ev_y = data[i, 1]
        gx = int(ev_x) // CELL_SIZE
        gy = int(ev_y) // CELL_SIZE
        
        if 0 <= gx < GRID_W and 0 <= gy < GRID_H:
            c = grid_counts[gx, gy]
            if c < MAX_EVENTS_PER_CELL:
                grid_indices[gx, gy, c] = i
                grid_counts[gx, gy] += 1

    return sigma_major_map, sigma_minor_map, sin_map, cos_map, norm_map, grid_counts, grid_indices

@njit(parallel=True, fastmath=True, cache=True)
def core_calculation(data, bkg_rate=1e-5, t_exp=1000.0):
    # 1. 预计算和构建网格
    s_maj, s_min, sa, ca, nc, g_cnt, g_idx = precompute_and_build_grid(data)
    
    ratio_grid = np.zeros((DATA_SIZE_X, DATA_SIZE_Y), dtype=np.float32)
    R_grid = np.zeros((DATA_SIZE_X, DATA_SIZE_Y), dtype=np.float32)

    psf_radius = 7.0
    iteration = 5
    bkg_rate_f = np.float32(bkg_rate)
    t_f = np.float32(t_exp)
    inv_bkg = 1.0 / bkg_rate_f

    # 2. 遍历 Grid (而不是遍历所有 Pixel)
    # prange 并行化 grid 遍历
    for gx in prange(GRID_W):
        for gy in range(GRID_H):
            # 检查当前 Grid 以及周围 3x3 邻域是否有光子
            # 如果邻域全是空的，这个 Grid 里的所有像素都不需要计算
            has_events = False
            for ix in range(max(0, gx-1), min(GRID_W, gx+2)):
                for iy in range(max(0, gy-1), min(GRID_H, gy+2)):
                    if g_cnt[ix, iy] > 0:
                        has_events = True
                        break
                if has_events: break
            
            if not has_events:
                continue

            # 3. 如果是活跃 Grid，遍历其中的像素
            x_start = gx * CELL_SIZE
            y_start = gy * CELL_SIZE
            # 边界保护
            x_end = min(x_start + CELL_SIZE, DATA_SIZE_X)
            y_end = min(y_start + CELL_SIZE, DATA_SIZE_Y)

            for x in range(x_start, x_end):
                for y in range(y_start, y_end):
                    
                    # --- Pixel Kernel Start ---
                    # 收集光子 (Gather)
                    # 只需搜索 3x3 的邻居格子
                    local_count = 0
                    local_s = np.zeros(256, dtype=np.float32) # Stack array

                    # Pixel geometric params
                    px_smaj = s_maj[x, y]
                    px_smin = s_min[x, y]
                    px_sa = sa[x, y]
                    px_ca = ca[x, y]
                    px_nc = nc[x, y]
                    
                    # 确定要搜索的 Grid 范围 (当前像素所属 Grid 的邻居)
                    # 注意：这里 x, y 属于 gx, gy，所以只需搜 gx, gy 的邻居
                    for ix in range(max(0, gx-1), min(GRID_W, gx+2)):
                        for iy in range(max(0, gy-1), min(GRID_H, gy+2)):
                            cnt = g_cnt[ix, iy]
                            if cnt == 0: continue
                            
                            for k in range(cnt):
                                idx = g_idx[ix, iy, k]
                                dx_ev = data[idx, 0] - x
                                dy_ev = data[idx, 1] - y
                                
                                # Bounding Box Check
                                if abs(dx_ev) <= psf_radius and abs(dy_ev) <= psf_radius:
                                    # PSF Calc
                                    t1 = (dx_ev * px_ca + dy_ev * px_sa) / px_smaj
                                    t2 = (-dx_ev * px_sa + dy_ev * px_ca) / px_smin
                                    expo = (t1*t1 + t2*t2) * 0.5
                                    val = math.exp(-expo) * px_nc
                                    
                                    if local_count < 256:
                                        local_s[local_count] = val
                                        local_count += 1
                    
                    if local_count == 0:
                        continue

                    # R Iteration
                    R = np.float32(0.05)
                    for _ in range(iteration):
                        temp_R = np.float32(0.0)
                        for k in range(local_count):
                            s = local_s[k]
                            temp_R += (R * s) / ((R * s + bkg_rate_f) * t_f)
                        R = temp_R
                    
                    R_grid[x, y] = R

                    # Log Likelihood
                    log_sum = np.float32(0.0)
                    for k in range(local_count):
                        # term = (R*s + bkg)/bkg = 1 + R*s*inv_bkg
                        log_sum += np.log(1.0 + R * local_s[k] * inv_bkg)
                    
                    ratio_grid[x, y] = log_sum - (t_f * R)
                    # --- Pixel Kernel End ---

    return ratio_grid, R_grid

@njit(fastmath=True, cache=True)
def find_local_maxima(ratio_grid, R_grid, threshold=8, window_size=21, border=10):
    """
    替代 scipy.ndimage.maximum_filter。
    只在大于阈值的点进行邻域检查，性能比 scipy 快得多。
    """
    w_radius = window_size // 2
    sources = []
    
    # 直接遍历，但加上步长或快速跳过小于阈值的区域
    # 这里为了简单直接全遍历，因为 Numba 循环极快
    rows, cols = ratio_grid.shape
    
    for r in range(border, rows - border):
        for c in range(border, cols - border):
            val = ratio_grid[r, c]
            
            if val < threshold:
                continue
            
            # 如果大于阈值，检查是否是局部最大
            is_max = True
            # 手动遍历窗口
            r_min = max(0, r - w_radius)
            r_max = min(rows, r + w_radius + 1)
            c_min = max(0, c - w_radius)
            c_max = min(cols, c + w_radius + 1)
            
            for i in range(r_min, r_max):
                for j in range(c_min, c_max):
                    if i == r and j == c:
                        continue
                    if ratio_grid[i, j] > val: # 严格大于，或者 >= 看来题意
                        is_max = False
                        break
                if not is_max:
                    break
            
            if is_max:
                # 记录结果: x, y, rate. 注意 data index 和坐标的对应
                # 原代码是 data[n, 0] 是 x, 对应 array 的 index 可能是 [x, y] 或 [y, x]
                # 假设 ratio_grid[x, y] 对应物理坐标 x, y
                sources.append((float(r), float(c), R_grid[r, c]))
                
    return sources

if __name__ == '__main__':
    t0 = time.time()

    # 读取数据
    hdu = fits.open('mock_data.fits', memmap=True)
    x_coord = hdu[1].data['x']
    y_coord = hdu[1].data['y']
    data = np.column_stack((x_coord, y_coord)).astype(np.float64)
    data = np.ascontiguousarray(data)

    t1 = time.time()
    print(f"Data IO time: {t1 - t0:.4f}s")

    # 计算
    print("Start calculating map ...")
    ratio_grid, R_grid = core_calculation(data)

    t2 = time.time()
    print(f"Map calculation time: {t2 - t1:.4f}s")

    # 检测
    source_coord_list = find_local_maxima(ratio_grid, R_grid)
    print(f"Detected sources: {len(source_coord_list)}")

    t3 = time.time()
    print(f"Detection time: {t3 - t2:.4f}s")

    # 保存结果
    source_coord_array = np.array(source_coord_list)
    # col1 = fits.Column(name='x', format='I', array=source_coord_array[:, 0])
    # col2 = fits.Column(name='y', format='I', array=source_coord_array[:, 1])
    # col3 = fits.Column(name='countrate', format='D', array=source_coord_array[:, 2])
    # cols = fits.ColDefs([col1, col2, col3])
    # tbhdu = fits.BinTableHDU.from_columns(cols)
    # prihdr = fits.Header()
    # prihdr['COMMENT'] = "This file storages the info of detected sources"
    # prihdu = fits.PrimaryHDU(header=prihdr)
    # hdulist = fits.HDUList([prihdu, tbhdu])
    # hdulist.writeto('detection_info.fits', overwrite=True)

    with fits.open('detection_info.fits', mode='update') as hdul:
        tb_hdu = hdul[1]
        tb_hdu.data['x'] = source_coord_array[:, 0]
        tb_hdu.data['y'] = source_coord_array[:, 1]
        tb_hdu.data['countrate'] = source_coord_array[:, 2]

    t4 = time.time()
    print(f"Data IO time: {t4 - t3:.4f}s")
    print(f"Total time: {t4 - t0:.4f}s")
